[Security] Configurable execution order for firewall listeners

scheb opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| License       | MIT
| Doc PR        | n/a

Hello there, I'm the author of `scheb/two-factor-bundle`, which extends Symfony's security layer with two-factor authentication. I've been closely following the recent changes by @wouterj to rework the security layer with "authenticators" (great work!). While I managed to make my bundle work with authenticators, I see some limitations in the security layer that I'd like to address to make such extensions easier to implement.

In #37336 I've submitted a draft to let security factories add their own authentication listeners to the firewall. This PR is intended to address the issue of execution order. If you look at the `Firewall` class

https://github.com/symfony/symfony/blob/f64f59a9c0d92fdd65f9de3e44b612402b224aaf/src/Symfony/Component/Security/Http/Firewall.php#L62-L82

authentication listeners are executed in the order of their creation. Additionally, there's hardcoded logic to execute `Symfony\Component\Security\Http\Firewall\AccessListener` always last and the logout listener second to last. I'd like to have a more flexible approach, to remove the hardcoded order and give authentication listeners the ability to determine their execution order.

I've added an optional interface to provide a priority to sort all registered authenitication listeners. Sorting is done in a compiler pass, so no time is wasted at runtime.

This is a draft, so I'd like to hear your opinion on this :)

scheb committed:
Add ability to prioritize firewall listeners

wouterj commented on src/Symfony/Component/Security/Http/Firewall.php:
```diff
@@ -59,26 +59,28 @@ public function onKernelRequest(RequestEvent $event)
         }
 
+        // Authentication listeners are pre-sorted by SortFirewallListenersPass
         $authenticationListeners = function () use ($authenticationListeners, $logoutListener) {
```
Can't we completely remove lines 63 - 84 if they are already pre-sorted?

scheb replied:
Unfortunately not, since you still need to yield `$logoutListener`, which is not included in the `$authenticationListeners` array, at the correct position.

I have the feeling this is a historic leftover to ensure execution order for the logout listener. I believe nowadays the logout listener could easily be added to `$authenticationListeners` array and let the priority take care of execution order. Then we could in fact remove all of these lines here.

Maybe someone can confirm the reasoning why the `$logoutListener` is not provided together with `$authenticationListeners`.

wouterj replied:
If you take a look at these two PRs, I think you're hypothesis is confirmed: https://github.com/symfony/symfony/pull/28100 + https://github.com/symfony/symfony/pull/28224

Maybe @chalasr can give a final confirmation though?

chalasr replied:
Correct, yes. 
For the record, the logout listener was originally moved outside of the `$authenticationListeners` list in https://github.com/symfony/symfony/pull/24805.
As long as we make sure that the logout and access listeners are executed last, there should be no issue for removing this reordering logic and relying on the priority-based order. 
Existing tests should make sure we don't break things anyways

wouterj commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/SortFirewallListenersPass.php:
```diff
@@ -0,0 +1,63 @@
+        }
+
+        return $priorities;
+    }
```
I think it's best to do the sorting directly in `SecurityExtension` - requiring #37336 to be merged first.

scheb replied:
That's what I initially thought, but I decided to go with a compiler pass for several reasons:

- The `SecurityExtension` is already very complex and I didn't want to make it even more complicated. The compiler pass allows for separation of concerns and easy testiability (tests not done yet).
- Doing the sorting late in the container building process via a compiler pass will also take dirty-injected services into account and sort them properly into order (like I do in: https://github.com/scheb/2fa/blob/ed2ce9804b6a78fe539894e77038ab96a52f5c56/src/bundle/DependencyInjection/Compiler/AccessListenerCompilerPass.php). If it's only sorted in `SecurityExtension` the correct sort order cannot be guaranteed.
- I have the feeling the container building must execute the "optimize" step first, before I can reliably fetch the class from the service definitions. Not sure about that though.

chalasr replied:
👍  The dedicated compiler pass makes sense here indeed, thanks for explaining.

scheb commented:
I've added the logout listener to the array of authentication listeners and removed the sorting logic from `Firewall`.

I have a bit of an issue with `CompleteConfigurationTest`. My compiler pass requires `ResolveChildDefinitionsPass` to be executed, but the test has disabled all optimization steps. The tests are failing because some `ChildDefintion`s are missing the `class` attribute, which would normally be set by `ResolveChildDefinitionsPass`. If I add `ResolveChildDefinitionsPass` as a optimization pass, other things are starting to break, because of missing child definitions.

I've tried to resolve the `class` recursively within my compiler pass (basically doing what `ResolveChildDefinitionsPass` would normally do for me) and then the tests go green. So it's an issue with test setup somehow. I don't really understand how to fix this.

wouterj commented:
When I make this change:

```diff
diff --git a/src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/CompleteConfigurationTest.php b/src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/CompleteConfigurationTest.php
index 95e3e6e345..226eb1a440 100644
--- a/src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/CompleteConfigurationTest.php
+++ b/src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/CompleteConfigurationTest.php
@@ -15,6 +15,7 @@ use PHPUnit\Framework\TestCase;
 use Symfony\Bundle\SecurityBundle\DependencyInjection\SecurityExtension;
 use Symfony\Bundle\SecurityBundle\SecurityBundle;
 use Symfony\Component\DependencyInjection\Argument\IteratorArgument;
+use Symfony\Component\DependencyInjection\Compiler\ResolveChildDefinitionsPass;
 use Symfony\Component\DependencyInjection\ContainerBuilder;
 use Symfony\Component\DependencyInjection\Reference;
 use Symfony\Component\Security\Core\Authorization\AccessDecisionManager;
@@ -69,10 +70,9 @@ abstract class CompleteConfigurationTest extends TestCase
         $configs = [];
         foreach (array_keys($arguments[1]->getValues()) as $contextId) {
             $contextDef = $container->getDefinition($contextId);
-            $arguments = $contextDef->getArguments();
-            $listeners[] = array_map('strval', $arguments['index_0']->getValues());
+            $listeners[] = array_map('strval', $contextDef->getArgument(0)->getValues());

-            $configDef = $container->getDefinition((string) $arguments['index_3']);
+            $configDef = $container->getDefinition((string) $contextDef->getArgument(3));
             $configs[] = array_values($configDef->getArguments());
         }

@@ -637,6 +637,7 @@ abstract class CompleteConfigurationTest extends TestCase

         $container = new ContainerBuilder();
         $container->setParameter('kernel.debug', false);
+        $container->register('cache.system', \stdClass::class);

         $security = new SecurityExtension();
         $container->registerExtension($security);
@@ -645,7 +646,7 @@ abstract class CompleteConfigurationTest extends TestCase
         $bundle->build($container); // Attach all default factories
         $this->getLoader($container)->load($file);

-        $container->getCompilerPassConfig()->setOptimizationPasses([]);
+        $container->getCompilerPassConfig()->setOptimizationPasses([new ResolveChildDefinitionsPass()]);
         $container->getCompilerPassConfig()->setRemovingPasses([]);
         $container->getCompilerPassConfig()->setAfterRemovingPasses([]);
         $container->compile();
```

There is only one failing test in the `CompleteConfigurationTest`:

```
1) Symfony\Bundle\SecurityBundle\Tests\DependencyInjection\XmlCompleteConfigurationTest::testFirewalls
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
         1 => 'security.user_checker'
         2 => '.security.request_matcher.xmi9dcw'
         3 => false
+        4 => false
+        5 => ''
+        6 => ''
+        7 => ''
+        8 => ''
+        9 => ''
+        10 => Array ()
+        11 => null
     )
     1 => Array (...)
     2 => Array (...)
     3 => Array (...)
 )

/home/wouter/symfony/src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/CompleteConfigurationTest.php:86
```

Seems like there are many new empty security listeners.

scheb commented:
@wouterj Nice, I'll try that and have a look at this failing test.

scheb commented on src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/CompleteConfigurationTest.php:
```diff
@@ -87,6 +87,14 @@ public function testFirewalls()
                 'security.user_checker',
                 '.security.request_matcher.xmi9dcw',
                 false,
+                false,
+                '',
+                '',
+                '',
+                '',
+                '',
+                [],
+                null
```
What's asserted here are the constructor arguments for the firewall's `FirewallConfig`. Now that `ResolveChildDefinitionsPass` is executed in the test, the default arguments provided by the `ChildDefiniton`s parent are propagated.

So I guess we just add them to the assert?

chalasr replied:
yes 👍

scheb replied:
Good, that's what I did :)

chalasr commented on src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/SecurityExtensionTest.php:
```diff
@@ -605,14 +606,15 @@ protected function getRawContainer()
     {
         $container = new ContainerBuilder();
         $container->setParameter('kernel.debug', false);
+        $container->register('cache.system', \stdClass::class);
```
This means that some service assume that `cache.system` is always available while it is not.
See #37362 which fixes the issue on lower branches, which should reduce the diff once merged up to master.

scheb commented:
So I tried to reproduce these failing tests now for an hour or so. I don't get it. I'm unable to reproduce these failures.

https://travis-ci.org/github/symfony/symfony/builds/700163046

These fail in `PhpCompleteConfigurationTest` because I have removed the code that adds logout listener to the authentication listeners array. So I expected these tests to fail.


But interestingly PHP: 7.3 + deps=high builds fine without any errors.

How's that possible?

wouterj commented:
Oh, looking at the composer logs I think `deps=high` always installs the master branch and not the PR branch. So that should be green once this PR is merged. I'm not totally sure though.

scheb commented:
https://travis-ci.org/github/symfony/symfony/builds/700383588

Squashed and rebased onto the bug #37362 fix. For some reason the failing tests are now working. Only failure comes from 
`ProcessTest::testUpdateStatus`, which is unrelated and looks like a fluke.

wouterj commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php:
```diff
@@ -285,6 +284,7 @@ private function createFirewalls(array $config, ContainerBuilder $container)
                 ->replaceArgument(1, $exceptionListener)
                 ->replaceArgument(2, $logoutListener)
                 ->replaceArgument(3, new Reference($configId))
+                ->addTag('security.firewall_map_context')
```
Since #37368 has been merged, there now is a `security.firewalls` parameter containing all firewall names. I propose to use that in the compiler pass instead of introducing a tag (see the referenced PR for a usage example).

Ryan tried to not introduce SecurityBundle "internal" tags, as this might confuse users. E.g. PHPstorm will auto complete this, so we should try to not clutter the autocompletion list with tags the user should never use. (in the end, we should maybe introduce the concept of internal tags?)

scheb replied:
Rebased and changed to use `security.firewalls` now.

scheb commented on src/Symfony/Component/Security/Http/composer.json:
```diff
@@ -31,7 +31,8 @@
         "symfony/event-dispatcher": "<4.3",
-        "symfony/security-csrf": "<4.4"
+        "symfony/security-csrf": "<4.4",
+        "symfony/security-bundle": "<5.2"
```
Have added this because both `symfony/security` and `symfony/security-bundle` need to be installed in the "=>5.2" version to work properly together.

chalasr replied:
`security-http` should not be aware of `security-bundle`, only security-bundle knows about security components and has version constraints for them

scheb replied:
Hmm, I'm wondering how would you ensure compatibility then? Combining a 5.1 `symfony/security-bundle` with 5.2 `symfony/security` would lead to an odd behavior.

wouterj commented on src/Symfony/Bundle/SecurityBundle/SecurityBundle.php:
```diff
@@ -75,6 +76,7 @@ public function build(ContainerBuilder $container)
         $container->addCompilerPass(new RegisterCsrfFeaturesPass());
         $container->addCompilerPass(new RegisterTokenUsageTrackingPass(), PassConfig::TYPE_BEFORE_OPTIMIZATION, 200);
         $container->addCompilerPass(new RegisterLdapLocatorPass());
+        $container->addCompilerPass(new SortFirewallListenersPass(), PassConfig::TYPE_BEFORE_REMOVING);
```
I think it's good to always add a comment when the priority is not the default. It is done with a reason, but that reason is often hidden and not visible when reading the code.
(I know that this is far from common in Symfony code, but I believe it's something Symfony source can improve a lot)

scheb replied:
Done :)

wouterj commented:
Hi @scheb! Is there anything missing from this PR, or can it be marked as ready-to-merge? I think it's a great feature to add to Symfony and feature freeze is only in a few weeks (so it's good to have a clear overview of what PRs can be ready before this deadline).

scheb commented:
@wouterj Technically it's ready and I'd like to see it in 5.2. It's still a draft, because I need feedback on this issue:

> Hmm, I'm wondering how would you ensure compatibility then? Combining a 5.1 `symfony/security-bundle` with 5.2 `symfony/security` would lead to an odd behavior.

It breaks the logout feature when you have such a combination in your application. And since @chalasr said

> `security-http` should not be aware of `security-bundle`, only `security-bundle` knows about security components and has version constraints for them

I'm not sure how we could solve this.

I tend to say the change to add the `LogoutListener` to the list of authentication listeners shouldn't be part of this PR. I added this part as an optimization, but it's not required to make the PR work. That specific change looks like a BC break to me, since it breaks compatibility among 5.x versions of `symfony/security-bundle` and `symfony/security`. So we're probably better off considering this change (to add the `LogoutListener` to the list of authentication) for Symfony 6.0.

What do you think?

fabpot commented:
What about removing the LogoutListener for now, merge this PR. Then, we create a PR for the LogoutListener and we can discuss the best option for BC. How does that sound?

scheb commented:
Hello there! I've removed the change regarding the LogoutListener, as Fabien suggested. The failed builds have been unrelated to the changes. So anything stopping up from merging the PR?

fabpot commented:
@scheb Thanks for the ping. Unfortunately, without a comment, there is no way I can see that something had changed here, except checking from time to time.

fabpot commented:
Thank you @scheb.

scheb commented:
@fabpot Thank you for merging! :)
